### ISSUE TITLE
Support text direction per line/paragraph

### ISSUE THREAD

**Issue by Mohamed-Hamza** [esm7/obsidian-rtl#18]
Hey good man,
Will u make it go RTL for only the scope (or the line) in which I make the RTL/LTR switch , not the whole note?
Mixing directions in the same note is needed please!! 

![image](https://user-images.githubusercontent.com/20577168/124360289-3374e200-dc29-11eb-8151-aec4a15deab9.png)

**Comment by Mohamed-Hamza** [comment]
and thanks man :)

**Comment by esm7** [comment]
I wish I could do that, but the underlying text editing engine (CodeMirror 5) only supports direction for the entire text at once.
Hopefully this will change in the near future with the next major version, CodeMirror 6, being more and more ready for prime time.

**Comment by zoli** [comment]
Now that https://github.com/codemirror/codemirror.next/issues/782 is closed I think we are closer to having per block direction.

**Comment by esm7** [comment]
That's amazing news, can't wait to try it out when Obsidian upgrades to this version!

**Comment by zoli** [comment]
I think per line RTL/LTR is now possible regarding that obsidian api has upgraded @codemirror/view v6.0.0.

https://github.com/obsidianmd/obsidian-api/issues/64#issuecomment-1196431428=

**Comment by zoli** [comment]
Maybe now we can reopen this issue?

**Comment by esm7** [comment]
It is indeed possible! I even did a quick dirty test and `perLineTextDirection` indeed works!
The thing is that this feature is not the same as 'auto' HTML direction. There is no automatic setting of the direction of any kind -- just the ability to specify different directions within the same document. So if I understand it correctly, to use this feature properly I'll need to add logic that decides on each line's direction and track it when the document changes.
Reopening the issue and will work on it at the closest opportunity.

**Comment by zoli** [comment]
> I'll need to add logic that decides on each line's direction and track it when the document changes.
You are right. I was thinking of STH like auto language detection of the line and using a map of languages which are RTL or LTR. Also a context menu for setting the direction manually would be nice.

**Comment by esm7** [comment]
Keeping metadata outside the note (e.g. manual direction) is a tricky business; This plugin does it but it works on a complete-note basis which is somehow still reasonable. But doing it on a per-line basis is probably very error prone, in the sense that if a file was modified outside of Obsidian (even due to syncing with a different device), the line numbers will mismatch and the wrong lines will be aligned.
I prefer a language detection approach similar to how 'auto' works for HTML, plus the ability to specify the default direction of a complete note.

**Comment by OranMizrahi** [comment]
Is there any way i cant do Mixed RTL&LTR same note

**Comment by esm7** [comment]
@OranMizrahi see the correspondence above. The basic ability was recently added to CodeMirror (the underlying editor that Obsidian uses), what's missing now is to write some logic around it in Obsidian.

**Comment by zoli** [comment]
Just wanted to note I'm working to implement this. [Here](https://github.com/zoli/obsidian-rtl/tree/feat/mixed-rtl-ltr) is the branch.

Most of live preview is done, But for the render part I should add a markdown post processor which I haven't started on that yet.

**Comment by esm7** [comment]
That's awesome and the choice of `franc` to power the logic is spot-on!
I only skimmed through the code and didn't fully review the flow, but one question that popped up for me is which events trigger `franc` on what input. We naturally wouldn't want to run it on the whole document for every key press. That's probably not the case right now, but I just wanted to make sure ahead of time :-)

**Comment by esm7** [comment]
I think that even if it gets to a point that it runs only on the current line, synchronously running the `franc` guesswork for every single key press is going to introduce noticeable latency, which will be disturbing for many users. (I personally suffer from visual latency introduced even by plugins that run regular expressions for triggering suggestion popups :( )
How about looking into the option of triggering it asynchronously, and only for the active line or paragraph?

**Comment by zoli** [comment]
> How about looking into the option of triggering it asynchronously, and only for the active line or paragraph?

Good idea I will look into it.

**Comment by zoli** [comment]
I was going to do a similar thing (checking the language of first word not the whole line) but then I saw that its possible to check the whole line without much problem and I remembered my problem with Gmail which was whenever I wanted to write Persian which is RTL but the first word was english I always had to change the sentence so the first word be Persian.
That's a problem which I as a Persian writer have with that type of implementation, Although it doesn't come up often.

The consistency with other environments that you mentioned is compelling enough for me to change the implementation to what you mentioned. 

Also even for franc I should've made an extra weight for starting word so it would took STH like `تست Test` that is near half english half persian as persian (it starts with persian word).

**Comment by zoli** [comment]
@esm7 I switched to checking first meaningful character and there is no need to `franc` anymore. Also while editing a file will only look into changing the view port event and changed line event, Still there is some shifting in the ranges because changing a line will change all next lines after range but no regex running on all next lines.
I've implemented the markdown post processor to, So I think there isn't much left. Will test it for couple of days and then make the pull request.
